### PR TITLE
Add secrets manager hooks

### PR DESCRIPTION
**Pull request by lox**
This adds a param `EnableSecretsManagerPlugin` that brings in the experimental https://github.com/buildkite/elastic-ci-stack-secrets-manager-hooks.

### CODE REVIEW

**Comment by lox** [comment]
Thoughts on how to bring this in @toolmantim? The idea is it would replace the s3 secrets stuff, but I'm unsure on how to approach deprecating the old stuff.

**Comment by toolmantim** [comment]
This is so great, much better than the S3 method.

I wonder if the S3 method is still best for general hook scripts support (maybe even adding more than just `environment` hooks as well) — and SM is for secrets only?

Also the other outstanding thing I was wondering was, how well would a end-user developer (or semi-experienced dev) go with setting a secret for their pipeline, if all we have is https://github.com/buildkite/elastic-ci-stack-secrets-manager-hooks? The readme at the moment covers a lot, and the prefix makes it tricky to know what to type exactly (vs a bucket, which is kind of just sitting there and has a web UI).

I think this method adds more onus on us to make it easy to use.

**Comment by toolmantim** [comment]
`gitcredential helper` 😍

**Comment by lox** [comment]
Yeah, it's tricky @toolmantim. Generally speaking, I think the developer experience for this is radically better, because you can set individual ENV values. The current setup you need to download an env file and modify it and re-upload it, with very specific SSE settings. 

Given users are several steps disconnected from the stack they are running on, I'm not sure how we give them exactly the correct syntax 🤔Same deal with the current s3 bucket deal though. Any ideas? 

I was thinking this morning that it would be neat to have the agent read it's entire config and hooks from secrets manager when it starts up. Then you'd have a central place to configure it.

File: templates/buildkite-elastic.yml [hunk 1]
```diff
@@ -438,6 +447,11 @@ Resources:
               - sns:Unsubscribe
               - sns:Subscribe
             Resource: "*"
+          - Effect: Allow
+            Action:
+              - secretsmanager:ListSecrets
+              - secretsmanager:GetSecretValue
+            Resource: "*"
```
Comment (toolmantim) [inline templates/buildkite-elastic.yml]: Can we limit this by prefix?

**Comment by toolmantim** [comment]
> The current setup you need to download an env file and modify it and re-upload it, with very specific SSE settings.

Yep totally. I'm more talking about the experience of figuring out how to do it for the very first time, for the person creating their very first stack. Perhaps someone who set up the elastic CI stack for the web UI.

The ongoing "maintain an env file" experience of S3 is definitely crap, it's just very accessible. But this is totally the better way, I think we might just need to help first-timers a bit more with docs.

Perhaps our SM examples should just assume the default stack name / default secrets prefix, e.g. remove all the `export secrets_manager_prefix="elastic-ci-stack"` lines from the readme.

Something like…

> ## Configuring Git Credentials
>
> Here's an example for how you'd configure a global git credential for the entire stack, using a [GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/):
>
> ```
> aws secretsmanager create-secret \
>   --name "buildkite-elastic-ci-stack/global/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```
>
> You can also override it a per-pipeline basis:
>
> ```
> aws secretsmanager create-secret \
>   --name "buildkite-elastic-ci-stack/pipeline/<pipeline-slug>/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```

Maybe we could just work to improve the readme together? Or should it go into elastic CI stack readme? 🤔

> I was thinking this morning that it would be neat to have the agent read it's entire config and hooks from secrets manager when it starts up. Then you'd have a central place to configure it.

Nice! The agent token could be read from SM. Do people use SM as a general config store?

**Comment by lox** [comment]
Perhaps we don't need that key prefix. I've been trying to figure out how teams will delegate access to things, perhaps we go with something like:

* `/buildkite/{org_slug}/ssh-private-key`
* `/buildkite/{org_slug}/git-credentials`
* `/buildkite/{org_slug}/env/{env_name}`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/ssh-private-key`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/git-credentials`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/env/{env_name}`

We could even add per-team support:

* `/buildkite/{org_slug}/team/{team_slug}/ssh-private-key`
* `/buildkite/{org_slug}/team/{team_slug}/git-credentials`
* `/buildkite/{org_slug}/team/{team_slug}/env/{env_name}`
* `/buildkite/{org_slug}/team/{team_slug}/pipeline/{pipeline_slug}/ssh-private-key`
* `/buildkite/{org_slug}/team/{team_slug}/pipeline/{pipeline_slug}/git-credentials`
* `/buildkite/{org_slug}/team/{team_slug}pipeline/{pipeline_slug}/env/{env_name}`

That would mean you structured your secrets by buildkite constructs. Would make it much easier to have "copy this command" things.

**Comment by lox** [comment]
We can do prefix matching on the secrets manager ARN's, so we would default to `/buildkite/*` but you could have a stack that only had access to particular org, or a set of pipelines 🤔

**Comment by toolmantim** [comment]
This seems really nice:

* `/buildkite/{org_slug}/ssh-private-key`
* `/buildkite/{org_slug}/git-credentials`
* `/buildkite/{org_slug}/env/{env_name}`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/ssh-private-key`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/git-credentials`
* `/buildkite/{org_slug}/pipeline/{pipeline_slug}/env/{env_name}`

You could also add:

* `/buildkite/agent-access-token`

**Comment by toolmantim** [comment]
Which would make it:

> ## Configuring Git Credentials
>
> Here's an example for how you'd configure a global git credential for the entire stack, using a [GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/):
>
> ```
> aws secretsmanager create-secret \
>   --name "/buildkite/<org-slug>/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```
>
> You can also override it a per-pipeline basis:
>
> ```
> aws secretsmanager create-secret \
>   --name "/buildkite/<org-slug>/pipeline/<pipeline-slug>/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```

I wonder if we should drop the initial `/` though?

> ## Configuring Git Credentials
>
> Here's an example for how you'd configure a global git credential for the entire stack, using a [GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/):
>
> ```
> aws secretsmanager create-secret \
>   --name "buildkite/<org-slug>/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```
>
> You can also override it a per-pipeline basis:
>
> ```
> aws secretsmanager create-secret \
>   --name "buildkite/<org-slug>/pipeline/<pipeline-slug>/git-credentials" \
>   --secret-string "https://<username>:<access-token>@github.com"
> ```

I think that seems nicer?

**Comment by lox** [comment]
> I wonder if we should drop the initial / though?

Yeah, that makes sense. I was following the convention that parameter store uses, but I don't think that is a thing with Secrets Manager.

It's really hard to know how to piece the things together Systems Manager Parameter Store is the other thing, which can store secrets too. There is a comparison [here](https://medium.com/@nimboya/aws-secrets-manager-vs-hashicorp-vault-vs-aws-parameter-store-bcbf60b0c0d1). The advantage the Secret Manager has is that it has a minimal web UI and it also supports key rotation (or at least makes it possible to write a lambda to do it).

**Comment by lox** [comment]
Cool. I reckon that is much nicer now.

**Comment by toolmantim** [comment]
Nice, I reckon so too!

I think adding access token support is probably really smart too… any security conscious person probably raises an eyebrow when pasting a token into a CloudFormation form field.

But if you start to push all the other bits of config into SM, you just gotta remember that each key costs (which is easier to swallow for a secret value)

**Comment by lox** [comment]
> /buildkite/agent-access-token

I was thinking that would be cool to actually have in the agent. It gets a bit blurrier how you map the many different types of agents you have onto which registration token.

**Comment by lox** [comment]
I was thinking of using https://aws.amazon.com/blogs/mt/integrating-aws-cloudformation-with-aws-systems-manager-parameter-store/ for the agent registration token in the cloudformation template to start with.

**Comment by lox** [comment]
Now the gnarly bit is how we replace the s3 stuff with the new secrets manager stuff, and how we document both in the Elastic Stack Readme.

**Comment by toolmantim** [comment]
Switch this to being the default, and have an option to enable "Legacy S3 Secrets Support" (which we drop in the future)?

**Comment by toolmantim** [comment]
(and that option can default to being on for now?)

**Comment by lox** [comment]
Yeah, that is the way I was leaning too. Then do we just link to https://github.com/buildkite/elastic-ci-stack-secrets-manager-hooks in the docs, or do we copy the docs into the main README?

File: templates/buildkite-elastic.yml [hunk 2]
```diff
@@ -449,9 +449,17 @@ Resources:
...
             Action:
-              - secretsmanager:ListSecrets
+              - secretsmanager:List*
+              - secretsmanager:Describe*
+            Resource: !Sub "arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:buildkite/*"
+          - Effect: Allow
+            Action:
               - secretsmanager:GetSecretValue
-            Resource: "*"
+            Resource: !Sub "arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:buildkite/*"
```
Comment (toolmantim) [inline templates/buildkite-elastic.yml]: Nice!

File: templates/buildkite-elastic.yml [hunk 3]
```diff
@@ -449,9 +449,17 @@ Resources:
...
+            Condition:
+              ForAnyValue:StringEquals:
+                "secretsmanager:VersionStage": "AWSCURRENT"
```
Comment (toolmantim) [inline templates/buildkite-elastic.yml]: Nice!

**Comment by toolmantim** [comment]
Looks like the `Configuration` section still needs to go: https://github.com/buildkite/elastic-ci-stack-secrets-manager-hooks#configuration

And then I think just link to that readme?

What did you think of my GitHub example for the `gitcredentials-helper`?

**Comment by lox** [comment]
> What did you think of my GitHub example for the gitcredentials-helper?

I thought it was good. Lemme wrangle it in there.

**Comment by toolmantim** [comment]
With the example, I figured if it was a global cred, then it wouldn't be limited to a GitHub org/repo because the access tokens aren't scoped like that either.

Related, should we make it clear that we don't offer any isolation of the secrets? In that, any pipeline's build can read secrets from anything in `buildkite/*`?

**Comment by lox** [comment]
README is updated now @toolmantim

**Comment by toolmantim** [comment]
> I've been pondering whether we could actually do better than this. We could accept a list of pipelines that the stack would work for as a CloudFormation param?

Interesting! Could make sense until agent queues can be locked down on the backend.

**Comment by lox** [comment]
Hmmmmm, maybe we should have used `buildkite/{org_slug}/queue/{queue_name}` as the prefix! 😅Then you could guarantee a stack could only access a give queue's secrets.

**Comment by lox** [comment]
Now that I think about it, it's kind of sketchy that different stacks can all access the entire `buildkite/*` secrets namespace. Makes it hard to create a `deploy` stack and a `default` stack with secrets and access control.

**Comment by toolmantim** [comment]
Oh yeah! 🤔 I guess using the queue name parameter in the permissions scope (and keyname) is probably the only viable way to do it?

**Comment by lox** [comment]
We could use tags? 🤔

```
# Create a secret available on all pipelines in a given queue
aws secretsmanager create-secret \
  --name "/buildkite/<org-slug>/git-credentials" \
  --secret-string "https://<username>:<access-token>@github.com" \
  --tags "Key=queue,Value=default"
```

**Comment by lox** [comment]
In the keyname is probably simpler. It would mean we'd need to bring back the org-slug in the CloudFormation template too, but that isn't a huge deal.

**Comment by toolmantim** [comment]
Oh interesting idea about tags! You could have the `default` queue not have the IAM permission to require the tag, but any other queue require the secret have the matching queue tag. So for the standard case, it's still simple.

**Comment by lox** [comment]
Although, the keys have to be unique, so you'd have collisions if you wanted different values for different key contexts.

**Comment by lox** [comment]
E.g have one `buildkite/<org-slug>/git-credentials` for your `default` queue and another value for your `deploy` queue.

**Comment by lox** [comment]
Unless we allowed `/buildkite/<org-slug>/git-credentials/1` etc

**Comment by toolmantim** [comment]
Yeah okay, I reckon you're right that just adding the queue name is probably the best way to do it. And those simple examples can just include `default` in them.

**Comment by lox** [comment]
I'm going to close this, it needs some more thought.

**Comment by lox** [comment]
How it will co-exist with the s3 secrets and how it fits into the release of 4.x vs perhaps going into the new ECS stack.

**Comment by lox** [comment]
Basically it was just too many variables for my brain to cope with 😅

**Comment by toolmantim** [comment]
Entirely fair!

**Comment by lox** [comment]
The other option would to do away with conventions for paths and just let folks provide an env var in their pipeline that specifies a Secrets Manager key to look up the ssh key or git-credentials for checkout. 

Something like `SSH_SECRETS_MANAGER_KEY` or `GIT_CREDENTIALS_SECRETS_MANAGER_KEY`. Then you would just need to make sure that your stack had access to what ever key that pointed to.

**Comment by lox** [comment]
Would end up being closer in spirit to https://github.com/buildkite/docker-ssh-env-config.

**Comment by toolmantim** [comment]
Do you mean, removing the pipeline slug part… so something like `buildkite/{queue_name}/{key}` with perhaps some magical env syntax? Like:

```
steps:
  - command: deploy.sh
    env:
      - SOME_SECRET="@some-secret-key"
```

and then that gets substituted via the SSM hook with the contents of `buildkite/{queue_name}/some-secret-key`, and can show a nice error if it doesn't exist (perhaps even showing an aws cli command example to set the key?

**Comment by toolmantim** [comment]
(I stole that syntax from https://zeit.co/docs/features/env-and-secrets btw)

**Comment by toolmantim** [comment]
(more docs here: https://zeit.co/docs/getting-started/secrets)

**Comment by lox** [comment]
The hard problem is getting ssh or https credentials for the initial checkout. That is 99% of the problem I'm trying to solve, the rest could be in a plugin with similar syntax to the above.

**Comment by toolmantim** [comment]
What's the problem with the original convention approach, but with only these two that are automatically searched for git creds:

* `buildkite/{queue_name}/{pipeline_slug}/ssh-private-key` (auto searched for via git checkout)
* `buildkite/{queue_name}/{pipeline_slug}/git-credentials` (auto searched for via git checkout)

If you want to use your own custom secrets, you use:

* `buildkite/{queue_name}/{pipeline_slug}/{key}`

```shell
aws secretsmanager create-secret \
  --name "buildkite/default/npm-token-ci-user" \
  --secret-string "..."
```

Then, if you wanted to expose that to every step in your pipeline:

```yml
env:
  - NPM_TOKEN="@npm-token-ci-user"
steps:
  - label: "🛠"
    command: test.sh
```

or just a single step:

```yml
steps:
  - label: "🛠"
    command: test.sh
    env:
      - NPM_TOKEN="@npm-token-ci-user"
```

or if you wanted to pass into a container:

```yml
steps:
  - label: "🛠"
    plugins:
      docker-compose#v2.5.1:
        run: tests
        env:
          - NPM_TOKEN
    env:
      - NPM_TOKEN="@npm-token-ci-user"
```

**Comment by lox** [comment]
@toolmantim yup, I think just implementing those two keys would be a pragmatic compromise. I'll do that. Thanks for talking it through.

**Comment by toolmantim** [comment]
👍🏼

I just updated the custom secret example, to remove the `pipeline_slug` part — I didnt want to give people the illusion that a secret is only visible to a pipeline, security wise.

But it made we wonder about the other guy creds ones.

How do we make it clear, in the API itself, that the secrets are visible to all pipelines? 🤔

**Comment by toolmantim** [comment]
Actually, I think the API we’ve got above is fine, as far as making that clear. We probably just need it littered through the readme and what-not too, that secrets are visible to jobs on that queue. And if you want to lock that to teams etc, you need a custom env hook (although we could add team-lockdown support as a stack param)

**Comment by lox** [comment]
@toolmantim my intention was to implement just the two automatically searched paths:

* buildkite/{queue_name}/{pipeline_slug}/ssh-private-key (auto searched for via git checkout)
* buildkite/{queue_name}/{pipeline_slug}/git-credentials (auto searched for via git checkout)

I think the other stuff is better handled holistically with a plugin or a top-level feature of the agent / platform.

**Comment by lox** [comment]
The reason I had the `{org_slug}` in there is that folks quite often have multiple orgs backed by the one AWS account. I think it's probably worth keeping in there. Thoughts @toolmantim?

**Pull request closed without merging**